File: lab1/sensor/src/main/java/edu/fer/rassus/sensor/networking/service/SensorService.java
package edu.fer.rassus.sensor.networking.service;

import edu.fer.rassus.sensor.model.Sensor;
import edu.fer.rassus.sensor.model.WeatherReading;
import io.reactivex.rxjava3.core.Single;
import retrofit2.adapter.rxjava3.Result;
import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Path;

public interface SensorService {

  @GET("sensor/nearest/{id}")
  Single<Result<Sensor>> findNearestSensor(@Path("id") int sensorId);

  @POST("sensor")
  Single<Result<Void>> registerSensor(@Body Sensor sensor);

  @POST("sensor/{id}")
  Single<Result<Void>> sendReading(
      @Path("id") int sensorId,
      @Body WeatherReading reading
  );

  @DELETE("sensor/{id}")
  Single<Result<Void>> unregisterSensor(@Path("id") int sensorId);
}

File: server/src/main/kotlin/edu/fer/rassus/server/service/SensorService.kt
package edu.fer.rassus.server.service

import edu.fer.rassus.server.AddSensorRequest
import edu.fer.rassus.server.model.Sensor
import edu.fer.rassus.server.repository.SensorRepository
import edu.fer.rassus.server.toEntity
import io.reactivex.Single
import org.springframework.stereotype.Service
import javax.transaction.Transactional

@Service
class SensorService(private val sensorRepository: SensorRepository) {

    @Transactional
    fun addSensor(body: AddSensorRequest): Single<Sensor> {
        return Single.create {
            val sensor = body.toEntity()
            it.onSuccess(sensorRepository.save(sensor))
        }
    }

    @Transactional
    fun findClosestTo(sensorId: Int): Single<Sensor> {
        return Single.create { subscriber ->
            val sensorOptional = sensorRepository.findById(sensorId)
            if (sensorOptional.isEmpty) {
                subscriber.onError(IllegalArgumentException("No sensor with id $sensorId"))
                return@create
            }

            val sensor = sensorOptional.get()
            val closesNeighbour = sensorRepository.findAll()
                .filter { it != sensor }
                .minByOrNull { sensor.distanceTo(it) }

            if (closesNeighbour == null) {
                subscriber.onError(IllegalArgumentException("No neighbours."))
                return@create
            }

            subscriber.onSuccess(closesNeighbour)
        }
    }

    @Transactional
    fun getAll(): Single<List<Sensor>> {
        return Single.create { it.onSuccess(sensorRepository.findAll()) }
    }

    @Transactional
    fun deleteSensor(sensorId: Int): Single<Boolean> {
        return Single.create {
            try {
                sensorRepository.deleteById(sensorId)
            } catch (e: Exception) {
                it.onSuccess(false)
                return@create
            }

            it.onSuccess(true)
        }
    }
}